#! /bin/sh /usr/share/dpatch/dpatch-run
## gcrypt-curl-squeeze.dpatch by Brian Kroth <bpkroth@gmail.com>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: Fix some threading issues that squeeze has between curl, gnutls, and gcrypt - RT #228055

@DPATCH@
diff -urNad '--exclude=CVS' '--exclude=.svn' '--exclude=.git' '--exclude=.arch' '--exclude=.hg' '--exclude=_darcs' '--exclude=.bzr' collectd-pw~/src/apache.c collectd-pw/src/apache.c
--- collectd-pw~/src/apache.c	2014-03-07 20:20:54.000000000 -0600
+++ collectd-pw/src/apache.c	2014-03-08 13:41:42.119752652 -0600
@@ -29,6 +29,16 @@
 #include "plugin.h"
 #include "configfile.h"
 
+#if HAVE_PTHREAD_H
+# include <pthread.h>
+#endif
+
+#if HAVE_GCRYPT_H
+# include <gcrypt.h>
+# include <errno.h>
+GCRY_THREAD_OPTION_PTHREAD_IMPL;
+#endif
+
 #include <curl/curl.h>
 
 enum server_enum
@@ -673,9 +683,12 @@
 
 static int apache_init (void) /* {{{ */
 {
+#if HAVE_GCRYPT_H
+	gcry_control (GCRYCTL_SET_THREAD_CBS, &gcry_threads_pthread);
+#endif
 	/* Call this while collectd is still single-threaded to avoid
 	 * initialization issues in libgcrypt. */
-	curl_global_init (CURL_GLOBAL_SSL);
+	curl_global_init (CURL_GLOBAL_ALL);
 	return (0);
 } /* }}} int apache_init */
 
diff -urNad '--exclude=CVS' '--exclude=.svn' '--exclude=.git' '--exclude=.arch' '--exclude=.hg' '--exclude=_darcs' '--exclude=.bzr' collectd-pw~/src/bind.c collectd-pw/src/bind.c
--- collectd-pw~/src/bind.c	2014-03-07 20:20:54.000000000 -0600
+++ collectd-pw/src/bind.c	2014-03-08 13:41:44.407793749 -0600
@@ -46,6 +46,16 @@
 # include <sys/select.h>
 #endif
 
+#if HAVE_PTHREAD_H
+# include <pthread.h>
+#endif
+
+#if HAVE_GCRYPT_H
+# include <gcrypt.h>
+# include <errno.h>
+GCRY_THREAD_OPTION_PTHREAD_IMPL;
+#endif
+
 #include <curl/curl.h>
 #include <libxml/parser.h>
 #include <libxml/xpath.h>
@@ -1386,9 +1396,13 @@
   if (curl != NULL)
     return (0);
 
+#if HAVE_GCRYPT_H
+  gcry_control (GCRYCTL_SET_THREAD_CBS, &gcry_threads_pthread);
+#endif
+
   /* Call this while collectd is still single-threaded to avoid
    * initialization issues in libgcrypt. */
-  curl_global_init (CURL_GLOBAL_SSL);
+  curl_global_init (CURL_GLOBAL_ALL);
 
   curl = curl_easy_init ();
   if (curl == NULL)
diff -urNad '--exclude=CVS' '--exclude=.svn' '--exclude=.git' '--exclude=.arch' '--exclude=.hg' '--exclude=_darcs' '--exclude=.bzr' collectd-pw~/src/curl.c collectd-pw/src/curl.c
--- collectd-pw~/src/curl.c	2014-03-07 20:20:54.000000000 -0600
+++ collectd-pw/src/curl.c	2014-03-08 13:41:29.467517494 -0600
@@ -28,6 +28,16 @@
 #include "utils_match.h"
 #include "utils_time.h"
 
+#if HAVE_PTHREAD_H
+# include <pthread.h>
+#endif
+
+#if HAVE_GCRYPT_H
+# include <gcrypt.h>
+# include <errno.h>
+GCRY_THREAD_OPTION_PTHREAD_IMPL;
+#endif
+
 #include <curl/curl.h>
 
 /*
@@ -569,7 +579,10 @@
     INFO ("curl plugin: No pages have been defined.");
     return (-1);
   }
-  curl_global_init (CURL_GLOBAL_SSL);
+#if HAVE_GCRYPT_H
+  gcry_control (GCRYCTL_SET_THREAD_CBS, &gcry_threads_pthread);
+#endif
+  curl_global_init (CURL_GLOBAL_ALL);
   return (0);
 } /* }}} int cc_init */
 
diff -urNad '--exclude=CVS' '--exclude=.svn' '--exclude=.git' '--exclude=.arch' '--exclude=.hg' '--exclude=_darcs' '--exclude=.bzr' collectd-pw~/src/curl_json.c collectd-pw/src/curl_json.c
--- collectd-pw~/src/curl_json.c	2014-03-07 20:20:54.000000000 -0600
+++ collectd-pw/src/curl_json.c	2014-03-08 13:41:33.895599901 -0600
@@ -32,6 +32,16 @@
 #include <sys/types.h>
 #include <sys/un.h>
 
+#if HAVE_PTHREAD_H
+# include <pthread.h>
+#endif
+
+#if HAVE_GCRYPT_H
+# include <gcrypt.h>
+# include <errno.h>
+GCRY_THREAD_OPTION_PTHREAD_IMPL;
+#endif
+
 #include <curl/curl.h>
 
 #include <yajl/yajl_parse.h>
@@ -958,9 +968,12 @@
 
 static int cj_init (void) /* {{{ */
 {
+#if HAVE_GCRYPT_H
+  gcry_control (GCRYCTL_SET_THREAD_CBS, &gcry_threads_pthread);
+#endif
   /* Call this while collectd is still single-threaded to avoid
    * initialization issues in libgcrypt. */
-  curl_global_init (CURL_GLOBAL_SSL);
+  curl_global_init (CURL_GLOBAL_ALL);
   return (0);
 } /* }}} int cj_init */
 
diff -urNad '--exclude=CVS' '--exclude=.svn' '--exclude=.git' '--exclude=.arch' '--exclude=.hg' '--exclude=_darcs' '--exclude=.bzr' collectd-pw~/src/curl_xml.c collectd-pw/src/curl_xml.c
--- collectd-pw~/src/curl_xml.c	2014-03-07 20:20:54.000000000 -0600
+++ collectd-pw/src/curl_xml.c	2014-03-08 13:41:36.787653022 -0600
@@ -30,6 +30,16 @@
 #include <libxml/xpath.h>
 #include <libxml/xpathInternals.h>
 
+#if HAVE_PTHREAD_H
+# include <pthread.h>
+#endif
+
+#if HAVE_GCRYPT_H
+# include <gcrypt.h>
+# include <errno.h>
+GCRY_THREAD_OPTION_PTHREAD_IMPL;
+#endif
+
 #include <curl/curl.h>
 
 #define CX_DEFAULT_HOST "localhost"
@@ -1034,9 +1044,12 @@
 
 static int cx_init (void) /* {{{ */
 {
+#if HAVE_GCRYPT_H
+  gcry_control (GCRYCTL_SET_THREAD_CBS, &gcry_threads_pthread);
+#endif
   /* Call this while collectd is still single-threaded to avoid
    * initialization issues in libgcrypt. */
-  curl_global_init (CURL_GLOBAL_SSL);
+  curl_global_init (CURL_GLOBAL_ALL);
   return (0);
 } /* }}} int cx_init */
 
diff -urNad '--exclude=CVS' '--exclude=.svn' '--exclude=.git' '--exclude=.arch' '--exclude=.hg' '--exclude=_darcs' '--exclude=.bzr' collectd-pw~/src/network.c collectd-pw/src/network.c
--- collectd-pw~/src/network.c	2014-03-07 20:20:54.000000000 -0600
+++ collectd-pw/src/network.c	2014-03-08 13:41:23.247403103 -0600
@@ -72,6 +72,7 @@
  */
 # define GCRYPT_NO_DEPRECATED
 # include <gcrypt.h>
+# include <errno.h>
 # if defined __APPLE__
 /* Re enable deprecation warnings */
 #  pragma GCC diagnostic warning "-Wdeprecated-declarations"
@@ -508,7 +509,7 @@
 
   gcry_check_version (NULL); /* before calling any other functions */
   gcry_control (GCRYCTL_SET_THREAD_CBS, &gcry_threads_pthread);
-  gcry_control (GCRYCTL_INIT_SECMEM, 32768);
+  gcry_control (GCRYCTL_DISABLE_SECMEM);
   gcry_control (GCRYCTL_INITIALIZATION_FINISHED);
 } /* }}} void network_init_gcrypt */
 
@@ -3393,6 +3394,9 @@
 	have_init = 1;
 
 #if HAVE_LIBGCRYPT
+	/* Make this behave a little more like gnutls_global_init() so that it
+	* matches the behavior of the curl plugins that have been patched. */
+	gcry_control (GCRYCTL_SET_THREAD_CBS, &gcry_threads_pthread);
 	network_init_gcrypt ();
 #endif
 
diff -urNad '--exclude=CVS' '--exclude=.svn' '--exclude=.git' '--exclude=.arch' '--exclude=.hg' '--exclude=_darcs' '--exclude=.bzr' collectd-pw~/src/write_http.c collectd-pw/src/write_http.c
--- collectd-pw~/src/write_http.c	2014-03-07 20:20:54.000000000 -0600
+++ collectd-pw/src/write_http.c	2014-03-08 13:41:39.443701816 -0600
@@ -34,6 +34,12 @@
 # include <pthread.h>
 #endif
 
+#if HAVE_GCRYPT_H
+# include <gcrypt.h>
+# include <errno.h>
+GCRY_THREAD_OPTION_PTHREAD_IMPL;
+#endif
+
 #include <curl/curl.h>
 
 /*
@@ -592,9 +598,12 @@
 
 static int wh_init (void) /* {{{ */
 {
+#if HAVE_GCRYPT_H
+  gcry_control (GCRYCTL_SET_THREAD_CBS, &gcry_threads_pthread);
+#endif
   /* Call this while collectd is still single-threaded to avoid
    * initialization issues in libgcrypt. */
-  curl_global_init (CURL_GLOBAL_SSL);
+  curl_global_init (CURL_GLOBAL_ALL);
   return (0);
 } /* }}} int wh_init */
 
